File: 프로그래머스/2/131701. 연속 부분 수열 합의 개수/연속 부분 수열 합의 개수.js
function solution(elements) {
    const circular = elements.concat(elements); // 배열을 두 배로 늘려 원형 수열 표현
    const sums = new Set(); // 중복 합 제거를 위한 Set 객체

    for (let i = 1; i <= elements.length; i++) { // 부분 수열의 길이 (1 ~ elements 길이)
        let sum = 0;
        
        for (let j = 0; j < elements.length; j++) { // 시작 인덱스
            if (j === 0) {
                // 첫 번째 길이에 대한 초기 합계 계산
                sum = circular.slice(0, i).reduce((acc, cur) => acc + cur, 0);
            } else {
                // 이전 합에서 맨 앞 값을 빼고, 새로운 맨 뒤 값을 더하여 최적화 (Sliding Window)
                sum = sum - circular[j - 1] + circular[j + i - 1];
            }
            sums.add(sum);
        }
    }

    return sums.size; // 고유한 합의 개수 반환
}

File: 프로그래머스/2/131701. 연속 부분 수열 합의 개수/연속 부분 수열 합의 개수.java
import java.util.HashSet;
import java.util.Set;

class Solution {
    public int solution(int[] elements) {
        Set<Integer> set = new HashSet<>();
        int n = elements.length;
        
        // 원형 수열 처리를 위해 기존 배열을 2배 크기로 확장
        int[] circularElements = new int[n * 2];
        for (int i = 0; i < n; i++) {
            circularElements[i] = elements[i];
            circularElements[i + n] = elements[i];
        }
        
        // 부분 수열의 길이 (1부터 n까지)
        for (int len = 1; len <= n; len++) {
            // 시작 인덱스 (0부터 n-1까지)
            for (int i = 0; i < n; i++) {
                int sum = 0;
                // 부분 수열의 합 계산
                for (int j = i; j < i + len; j++) {
                    sum += circularElements[j];
                }
                set.add(sum);
            }
        }
        
        return set.size();
    }
}
